dcc-portal-ui/app/scripts/oncogrid/js/controller.js

```diff
@@ -113,4 +113,157 @@
         };
 
     });
+
+    module.controller('OncoGridUploadController', function(Restangular, SetNameService, donorsLimit, genesLimit, filters, 
+        LocationService, $scope, $modalInstance, $location, $q, SetService, $timeout){
+
+        $scope.donorsLimit = donorsLimit;
+        $scope.genesLimit = genesLimit;
+        $scope.filters = filters;
+        $scope.isLaunchingOncoGrid = false;
+
+        $scope.maxDonorsLimit = 3000;
+        $scope.maxGenesLimit = 100;
+
+        $scope.params = {};
+        $scope.hasValidParams= false;
+
+        $scope.params.donorsCount = Math.min($scope.donorsLimit || 3000, $scope.maxDonorsLimit);
+        $scope.params.genesCount = Math.min($scope.genesLimit || 100, $scope.maxGenesLimit);
+        $scope.params.setName = '';
+
+        $scope.hasValidDonorCount = function(value){
+            var val = parseInt(value,10);
+            if (isNaN(val)) {
+                return false;
+            }
+            if (!angular.isNumber(val) || val > $scope.maxDonorsLimit || val <= 0 || val > $scope.donorsLimit) {
+                return false;
+            }
+            return true;
+        }
+
+        $scope.hasValidGeneCount = function(value){
+            var val = parseInt(value,10);
+            if (isNaN(val)) {
+                return false;
+            }
+            if (!angular.isNumber(val) || val > $scope.maxGenesLimit || val <= 0 || val > $scope.genesLimit) {
+                return false;
+            }
+            return true;
+        }
+
+        $scope.checkInput = function() {
+            var params = $scope.params;
+            if ($scope.hasValidDonorCount(params.donorsCount) && 
+                $scope.hasValidGeneCount(params.genesCount)) {
+                $scope.hasValidParams = true;
+            } else {
+                $scope.hasValidParams = false;
+            }
+        };
+
+        $scope.getSetName = function(filters){
+            return SetNameService.getSetFilters()
+                .then(function (filters) {
+                    return SetNameService.getSetName(filters);
+                })
+                .then(function (setName) {
+                    $scope.params.setName = setName;
+                });
+        }
+
+        $scope.getDonorsParams = function(){
+
+            var donorSetParams = {
+                filters: $scope.filters || {},
+                size: $scope.params.donorsCount,
+                type: 'donor',
+                isTransient: true,
+                name: `Top ${$scope.params.donorsCount} Donors: ${_.includes($scope.params.setName, 'All') ? '' : $scope.params.setName}`
+            };
+
+            return donorSetParams;
+        }
+
+        $scope.getGenesParams = function(){
+
+             var geneSetParams = {
```

Comment:
This is not what I meant, but even the above has duplication and can be improved:

```js
const getSetParams = (entity) => ({
    const size  = entity === 'donor' ? $scope.params.donorsCount : $scope.params.genesCount;
    filters: $scope.filters || {},
    size: size,
    type: entity,
    isTransient: true,
    name: `Top ${size} ${_.capitalize(entity)}s: ${_.includes($scope.params.setName, 'All') ? '' : $scope.params.setName}`
})
```

What I was suggesting was to move out the ternary:


```js
const getSetParams = (entity, count) => ({
    filters: $scope.filters || {},
    size: count,
    type: entity,
    isTransient: true,
    name: `Top count ${_.capitalize(entity)}s: ${_.includes($scope.params.setName, 'All') ? '' : $scope.params.setName}`
})
```

This is now distilled. If you wish, you could create wrapper methods to bind in the scope variable if it is used more than a few times.